pkg/ignore/types.go

```diff
@@ -0,0 +1,108 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ignore
+
+import (
+	"os"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// Ignorer  helps identifying if a BundleFile needs to be ignored
+type Ignorer interface {
+	// Ignore checks file has to be ignored or not, returns true if it needs to be ignored
+	Ignore(path string, fi os.FileInfo) Ignoreable
+	// Patterns returns the pattern associated with interface
+	Patterns() *sets.String
+}
+
+// IgnoredType helps classifying the ignored file.
+// This helps in WalkFunc to do filepath.SkipDir for more efficient traversals
+type IgnoredType string
+
+const (
+
+	// Current holds all direct files that are ignored
+	Current IgnoredType = "Current"
+
+	// Transitive holds all the directory types that are ignored
+	// these types might have directory ignored, but can have invert patterns
+	Transitive IgnoredType = "Transitive"
+)
+
+// Ignoreable holds whether the file or directory is qualified to be ignored
+type Ignoreable struct {
+	Ignored bool
+	Type    IgnoredType
+}
+
+var _ Ignorer = (*defaultIgnorer)(nil)
+
+// Transitive  represents the file that will be added to the bundle
+type bundleFile struct {
+	// rootDir the root directory where the .dockerignore is searched for
+	rootDir string
+	// name the basename of the file that needs to be checked for ignorability
+	name string
+	// path the absolute path of the file that needs to be checked for ignorability
+	path string
+	// isDir the file that is checked for ignorability is a directory ?
+	isDir bool
+	// patterns the patterns that needs to be checked with this BundleFile
+	patterns *sets.String
+}
+
+// defaultIgnorer is the default Ignorer which is returned when no .dockerignore file is present
+// or error processing .dockerignore
+type defaultIgnorer struct {
+	patterns *sets.String
+}
+
+// DefaultPatterns  holds the list of patterns that are always ignored
+var DefaultPatterns = []string{".git", "vendor", "node_modules"}
+
+// Ignore implements Ignorer, for no dockerignore cases, where only .git is ignored
+func (i *defaultIgnorer) Ignore(path string, fi os.FileInfo) Ignoreable {
+	bundleFile := &bundleFile{
+		name:     fi.Name(),
+		path:     path,
+		rootDir:  directory,
+		isDir:    fi.IsDir(),
+		patterns: i.patterns,
+	}
+
+	ignorable := &Ignoreable{}
+
+	if directory != path {
+		if !checkIfParentIncluded(path, fi) {
+			ignorable.Ignored = bundleFile.ignore()
+			if fi.IsDir() {
+				ignorable.Type = Transitive
+			} else {
+				ignorable.Type = Current
+			}
+
+		}
+	}
```

Comment:
This feels like overkill for the `defaultIgnorer` implementation.  I was literally thinking it would be something along the lines of what we had before...
```go
if fi.IsDir() {
  if path == ".git" || path == "vendor" ... {
    return Transitive
  }
} else if  path == "blah" {
  return Current
}
return NotIgnored
```

Reply:
I have given an different impl in mu current WIP